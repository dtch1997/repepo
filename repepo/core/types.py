from typing import List, NamedTuple, Union

from transformers import PreTrainedTokenizerBase, PreTrainedModel
from repepo.repe.rep_control_reading_vec import WrappedReadingVecModel

# Placeholder type definitions
Model = Union[PreTrainedModel, WrappedReadingVecModel]
Tokenizer = PreTrainedTokenizerBase


# Base types
class Example(NamedTuple):
    instruction: str
    input: str
    output: str


class RepExample(NamedTuple):
    instruction: str 
    input: str
    output: str # for repe, we do not need an ouput, but this is not necessarily going to be the same as direction. Eg. we could have an example for 'the amount of bias is' and the output could be 'high' or 'low', not True or False
    id: int # for repe reading, we have two examples per id, one for each direction
    direction: bool # True for positive, False for negative


class Completion(NamedTuple):
    prompt: str
    response: str


Dataset = Union[List[Example], List[RepExample]]
