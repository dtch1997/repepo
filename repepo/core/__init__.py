from .types import BaseDataset
from .types import BasePipeline
