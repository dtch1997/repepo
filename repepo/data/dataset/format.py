import abc 
import random

from typing import Dict, Any, NewType, List

Example = NewType('Example', Dict[str, Any])
# Example: 
#   - a dict with keys 'instruction', 'input', 'output'
#   - could also have additional metadata

Completion = NewType('Completion', Dict[str, Any])
# Completion: 
#   - a dict with keys 'prompt', 'response'

def completion_to_str(completion: Completion) -> str:
    return completion['prompt'] + completion['response']

class AbstractFormatter(abc.ABC):
        
    def format_example(self, example: Example, **kwargs) -> Completion:
        pass

    def apply(self, examples: List[Example]) -> List[Completion]:
        completions = []
        for example in examples:
            completion = self.format_example(example)
            completions.append(completion)
        return completions
    
class QAFormatter(AbstractFormatter):
    """ Format as a simple QA pair. """
    QUESTION = "Q: {instruction} {input}"
    ANSWER = "A: {output}"

    def format_example(self, example: Dict[str, Any], **kwargs):
        del kwargs
        prompt = self.QUESTION.format_map(example)
        response = self.ANSWER.format_map(example)
        return {
            'prompt': prompt, 
            'response': response
        }


class InstructionFormatter(AbstractFormatter):
    """ Instruction formatter used for fine-tuning Alpaca. """

    PROMPT_INPUT: str = (
        "Below is an instruction that describes a task, paired with an input that provides further context. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:"
    )
    PROMPT_NO_INPUT: str = (
        "Below is an instruction that describes a task. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Response:"
    )

    def format_example(self, example: Dict[str, Any], **kwargs):
        del kwargs
        if 'input' in example and bool(example['input']):
            prompt = self.PROMPT_INPUT.format_map(example)
        else:
            prompt = self.PROMPT_NO_INPUT.format_map(example)
        response = example['output']
        return {
            'prompt': prompt, 
            'response': response
        }

class FewShotPrompter:
    """ Compose examples few-shot """
    
    def __init__(self, n_few_shot_examples: int = 2):
        self.n_few_shot_examples = n_few_shot_examples
    
    def apply(self, completions: List[Completion]) -> List[Completion]:

        output_completions = []

        for i, completion in enumerate(completions):

            # Sample different completions for context
            few_shot_examples = []
            selected_idxes = [i, ]        
            for _ in range(self.n_few_shot_examples):
                done = False
                while not done:
                    idx = random.randint(0, len(completions) - 1) # range inclusive
                    done = idx not in selected_idxes
                few_shot_examples.append(
                    completion_to_str(completions[idx])
                )
                selected_idxes.append(idx)

            # Concatenate completions
            prompt = '\n'.join(few_shot_examples) + completion['prompt']
            response = completion['response']
            output_completions.append(dict(
                prompt=prompt, 
                response=response
            ))

        return output_completions