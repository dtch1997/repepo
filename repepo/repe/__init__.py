import warnings

warnings.filterwarnings("ignore")


from .pipelines import repe_pipeline_registry

# RepControl
from .rep_control_pipeline import *

# RepReading
from .rep_readers import *
from .rep_reading_pipeline import *
