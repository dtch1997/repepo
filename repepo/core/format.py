import abc
from typing import List

from repepo.core.types import Completion
from repepo.core.types import Example


class AbstractFormatter(abc.ABC):
    """Describes how to format examples as completions"""

    @abc.abstractmethod
    def apply(self, example: Example, **kwargs) -> str:
        raise NotImplementedError()

    def apply_list(self, examples: List[Example]) -> List[Completion]:
        completions = []
        for example in examples:
            completion = self.apply(example)
            completions.append(completion)
        return completions


class InputOutputFormatter(AbstractFormatter):
    """Format as a simple input-output pair."""

    PROMPT_TEMPLATE = "Input: {instruction} {input} \n" "Output: "

    def apply(self, example: Example, **kwargs):
        del kwargs
        return Completion(
            prompt=self.PROMPT_TEMPLATE.format(
                instruction=example.instruction, input=example.input
            ),
            response=example.output,
        )


class InstructionFormatter(AbstractFormatter):
    """Instruction formatter used for fine-tuning Alpaca."""

    PROMPT_INPUT: str = (
        "Below is an instruction that describes a task, paired with an input that provides further context. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Input:\n{input}\n\n### Response:"
    )
    PROMPT_NO_INPUT: str = (
        "Below is an instruction that describes a task. "
        "Write a response that appropriately completes the request.\n\n"
        "### Instruction:\n{instruction}\n\n### Response:"
    )

    def apply(self, example: Example, **kwargs):
        del kwargs
        if bool(example.input):
            prompt = self.PROMPT_INPUT.format(
                instruction=example.instruction, input=example.input
            )
        else:
            prompt = self.PROMPT_NO_INPUT.format_map(
                {"instruction": example.instruction}
            )
        response = example.output
        return Completion(prompt=prompt, response=response)
