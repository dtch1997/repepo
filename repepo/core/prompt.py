import abc
from dataclasses import dataclass
from typing import List
from typing_extensions import override

from repepo.core.types import Completion


def completion_to_str(completion: Completion) -> str:
    return completion.prompt.rstrip() + " " + completion.response.lstrip() + "\n"


class Prompter(abc.ABC):
    """Interface for modifying completions"""

    @abc.abstractmethod
    def apply(self, completion: Completion) -> Completion:
        raise NotImplementedError()

    def apply_list(self, completions: List[Completion]) -> List[Completion]:
        return [self.apply(completion) for completion in completions]


class IdentityPrompter(Prompter):
    """Return the prompt as-is"""

    @override
    def apply(self, completion):
        return completion


@dataclass
class LlamaChatPrompter(Prompter):
    """
    Prepend system message before the first prompt. This is based on the SycophancySteering
    prompt format, except the system message is not inside of the [INST] tags. Hopefully
    this is OK?

    TODO: Put the sys message inside of the [INST] tags, or at least make it possible to do so.
          This will require some rearchitecting

    Based on: https://github.com/nrimsky/SycophancySteering/blob/main/utils/tokenize_llama.py#L30
    """

    system_prompt: str = "You are a helpful, honest and concise assistant."

    B_SYS = "<<SYS>>\n"
    E_SYS = "\n<</SYS>>\n\n"

    @override
    def apply(self, completion: Completion) -> Completion:
        prompt = f"{self.B_SYS}{self.system_prompt}{self.E_SYS}{completion.prompt}"
        response = completion.response
        return Completion(prompt=prompt, response=response)


class FewShotPrompter(Prompter):
    """Compose examples few-shot"""

    # TODO: should these be randomized?
    few_shot_completions: list[Completion]

    def __init__(self, few_shot_completions: list[Completion]):
        self.few_shot_completions = few_shot_completions

    @override
    def apply(self, completion: Completion) -> Completion:
        prompt = ""
        for fs_comp in self.few_shot_completions:
            prompt += completion_to_str(fs_comp) + "\n"
        prompt += completion.prompt

        response = completion.response

        return Completion(prompt=prompt, response=response)
