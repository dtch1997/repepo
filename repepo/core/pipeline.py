from dataclasses import dataclass, field
from typing import Any, Optional

from transformers import GenerationConfig

from .types import Example, Model, Tokenizer
from .prompt import AbstractPrompter, IdentityPrompter
from .format import AbstractFormatter, InputOutputFormatter


@dataclass
class Pipeline:
    """Generation pipeline"""

    model: Model
    tokenizer: Tokenizer
    prompter: AbstractPrompter = field(default_factory=IdentityPrompter)
    formatter: AbstractFormatter = field(default_factory=InputOutputFormatter)

    def build_generation_prompt(self, example: Example) -> str:
        """Build a prompt for generation"""
        completion = self.formatter.apply(example)
        completion = self.prompter.apply(completion)
        return completion.prompt

    def generate(
        self,
        example: Example,
        config: Optional[GenerationConfig] = None,
        remove_base_prompt: bool = True,
    ) -> str:
        """Generate a completion for a given example"""
        base_prompt = self.build_generation_prompt(example)
        inputs: Any = self.tokenizer(base_prompt, return_tensors="pt")
        inputs = inputs.to(self.model.device)
        outputs = self.model.generate(
            **inputs,
            config=config,
        )[0]
        outputs_str = self.tokenizer.decode(outputs, skip_special_tokens=True)
        if remove_base_prompt:
            return outputs_str.replace(base_prompt, "")
        return outputs_str
