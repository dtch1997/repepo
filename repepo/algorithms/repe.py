from contextlib import contextmanager
from dataclasses import replace
from typing import Literal, Optional
from typing_extensions import override
import random
from repepo.core.pipeline import PipelineContext

from steering_vectors import (
    LayerType,
    SteeringVector,
    train_steering_vector,
    SteeringVectorTrainingSample,
    LayerMatcher,
    ModelLayerConfig,
    guess_and_enhance_layer_config,
)

from repepo.core import Pipeline
from repepo.core.format import Formatter
from repepo.core.types import Dataset, Example, Model, Tokenizer
from repepo.algorithms.base import Algorithm

MultiAnswerMethod = Literal["first_incorrect", "random_incorrect", "repeat_correct"]

DEFAULT_READING_TEMPLATE = "Consider the amount of truthfulness in the following answer.\n\n{question} {answer}\n\nThe amount of truthfulness in the answer is "


def _validate_reading_template(reading_template: str):
    if "{answer}" not in reading_template:
        raise ValueError(
            "reading_template must contain {answer} to be used with RepEngReadingControl"
        )
    if "{question}" not in reading_template:
        raise ValueError(
            "reading_template must contain {question} to be used with RepEngReadingControl"
        )


class RepeReadingControl(Algorithm):
    layer_type: LayerType
    multi_answer_method: MultiAnswerMethod
    reading_template: str
    layers: list[int] | None
    direction_multiplier: float
    layer_config: ModelLayerConfig | None
    patch_generation_tokens_only: bool
    seed: int

    def __init__(
        self,
        reading_template: str = DEFAULT_READING_TEMPLATE,
        layer_type: LayerType = "decoder_block",
        multi_answer_method: MultiAnswerMethod = "first_incorrect",
        seed: int = 0,
        layers: Optional[list[int]] = None,
        layer_config: Optional[ModelLayerConfig] = None,
        direction_multiplier: float = 1.0,
        patch_generation_tokens_only: bool = True,
        skip_reading: bool = False,
        override_vector: SteeringVector | None = None,
        skip_control: bool = False,
    ):
        self.multi_answer_method = multi_answer_method
        self.layer_type = layer_type
        self.seed = seed
        self.patch_generation_tokens_only = patch_generation_tokens_only
        _validate_reading_template(reading_template)
        self.reading_template = reading_template
        self.layers = layers
        self.layer_config = layer_config
        self.direction_multiplier = direction_multiplier

        self.skip_reading = skip_reading
        self.override_vector = override_vector
        self.skip_control = skip_control
        if self.skip_reading and self.override_vector is None:
            raise RuntimeError(
                "Either reading or override vector must be provided for control"
            )

    def _build_steering_vector_training_data(
        self, dataset: Dataset, formatter: Formatter
    ) -> list[SteeringVectorTrainingSample]:
        paired_prompts: list[SteeringVectorTrainingSample] = []
        for example in dataset:
            example_prompts = self._convert_example_to_training_samples(
                example, formatter
            )
            paired_prompts.extend(example_prompts)
        return paired_prompts

    def _get_layer_matcher_for_model(self, model: Model) -> LayerMatcher:
        layer_config = guess_and_enhance_layer_config(model, self.layer_config)
        if self.layer_type not in layer_config:
            raise ValueError(
                f"layer_type {self.layer_type} not found in model layer_config, please provide an expicit layer_config"
            )
        return layer_config[self.layer_type]

    def _get_steering_vector(
        self, pipeline: Pipeline, dataset: Dataset
    ) -> SteeringVector:
        repe_training_data = self._build_steering_vector_training_data(
            dataset, pipeline.formatter
        )
        return train_steering_vector(
            pipeline.model,
            pipeline.tokenizer,
            repe_training_data,
            layers=self.layers,
            layer_type=self.layer_type,
            layer_config=self.layer_config,
        )

    @override
    def run(self, pipeline: Pipeline, dataset: Dataset) -> Pipeline:
        # Steering vector reading
        # NOTE: The hooks read from this steering vector.

        if self.override_vector is not None:
            steering_vector: SteeringVector = self.override_vector
        elif self.skip_reading:
            steering_vector: SteeringVector = self._get_steering_vector(
                pipeline, dataset
            )
        else:
            raise RuntimeError(
                "Either reading or override vector must be provided for control"
            )

        # Creating the hooks that will do steering vector control
        # NOTE: How this works is that we create a context manager that creates a hook
        # whenever we are in a `PipelineContext`'s scope.
        # After exiting the context, the hook is deleted.

        # The PipelineContext is created in both `pipeline.generate` or `pipeline.calculate_output_logprobs`

        # need to use a hook so we can inspect the current thing being generated to know
        # which tokens to patch
        @contextmanager
        def steering_hook(context: PipelineContext):
            handle = None
            try:
                min_token_index = 0
                if self.patch_generation_tokens_only:
                    min_token_index = _find_generation_start_token_index(
                        pipeline.tokenizer,
                        context.base_prompt,
                        context.full_prompt,
                    )
                handle = steering_vector.patch_activations(
                    model=pipeline.model,
                    layer_config=self.layer_config,
                    # NOTE: if the direction multiplier is changed,
                    # subsequent generations will use the new value
                    # because this is a reference to the outer scope.
                    # This is probably counterintuitive
                    # NOTE: Same goes for layer_config above,
                    # but this is less critical because layer config is likely static
                    # TODO: change at some point.
                    multiplier=self.direction_multiplier,
                    min_token_index=min_token_index,
                )
                yield
            finally:
                if handle is not None:
                    handle.remove()

        if not self.skip_control:
            pipeline.hooks.append(steering_hook)

        return pipeline

    def _convert_example_to_training_samples(
        self, example: Example, formatter: Formatter
    ) -> list[SteeringVectorTrainingSample]:
        """Converts an example to the format expected by steering-vectors"""
        if example.incorrect_outputs is None:
            raise ValueError("RepeReadingControl requires incorrect_outputs to be set")
        incorrect_examples = [
            replace(example, output=output) for output in example.incorrect_outputs
        ]
        correct_examples = [example]
        if self.multi_answer_method == "repeat_correct":
            correct_examples = [example] * len(example.incorrect_outputs)
        elif self.multi_answer_method == "first_incorrect":
            incorrect_examples = [incorrect_examples[0]]
        elif self.multi_answer_method == "random_incorrect":
            rand_gen = random.Random(f"{self.seed}-{example.input}")
            incorrect_examples = [rand_gen.choice(incorrect_examples)]
        else:
            raise ValueError(f"Unknown multi_answer_method {self.multi_answer_method}")
        assert len(incorrect_examples) == len(correct_examples)
        paired_completions = [
            (formatter.apply(pos), formatter.apply(neg))
            for pos, neg in zip(correct_examples, incorrect_examples)
        ]
        return [
            SteeringVectorTrainingSample(
                positive_prompt=self.reading_template.format(
                    question=pos.prompt, answer=pos.response
                ),
                negative_prompt=self.reading_template.format(
                    question=neg.prompt, answer=neg.response
                ),
            )
            for pos, neg in paired_completions
        ]


def _find_generation_start_token_index(
    tokenizer: Tokenizer, base_prompt: str, full_prompt: str
) -> int:
    """Finds the index of the first generation token in the prompt"""
    base_toks = tokenizer.encode(base_prompt)
    full_toks = tokenizer.encode(full_prompt)
    prompt_len = len(base_toks)
    # try to account for cases where the final prompt token is different
    # from the first generation token, ususally weirdness with spaces / special chars
    for i, (base_tok, full_tok) in enumerate(zip(base_toks, full_toks)):
        if base_tok != full_tok:
            prompt_len = i
            break
    # The output of the last prompt token is the first generation token
    # so need to subtract 1 here
    return prompt_len - 1
