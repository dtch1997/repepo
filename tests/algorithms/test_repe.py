import torch
from repepo.algorithms.repe import (
    RepeReadingControl,
    _find_generation_start_token_index,
)
from repepo.core.format import InputOutputFormatter, LlamaChatFormatter
from repepo.core.types import Dataset, Example, Tokenizer
from repepo.core.pipeline import Pipeline
from repepo.core.prompt import LlamaChatPrompter
from syrupy import SnapshotAssertion
from transformers import GPTNeoXForCausalLM, LlamaForCausalLM
from tests._original_caa.llama_wrapper import LlamaWrapper


def test_RepeReadingControl_build_steering_vector_training_data_picks_one_neg_by_default(
    snapshot: SnapshotAssertion,
) -> None:
    dataset: Dataset = [
        Example(
            instruction="",
            input="Paris is in",
            output="France",
            incorrect_outputs=["Germany", "Italy"],
        ),
        Example(
            instruction="",
            input="1 + 1 =",
            output="2",
            incorrect_outputs=["11", "34", "3.14"],
        ),
    ]
    formatter = InputOutputFormatter()
    algorithm = RepeReadingControl()
    training_data = algorithm._build_steering_vector_training_data(dataset, formatter)
    assert len(training_data) == 2
    # should pick the first incorrect output only by default
    assert "France" in training_data[0].positive_prompt
    assert "Germany" in training_data[0].negative_prompt
    assert "2" in training_data[1].positive_prompt
    assert "11" in training_data[1].negative_prompt
    assert training_data == snapshot


def test_RepeReadingControl_build_steering_vector_training_data_with_random_incorrect() -> (
    None
):
    dataset: Dataset = [
        Example(
            instruction="",
            input="Paris is in",
            output="France",
            incorrect_outputs=["Germany", "Italy"],
        ),
        Example(
            instruction="",
            input="1 + 1 =",
            output="2",
            incorrect_outputs=["11", "34", "3.14"],
        ),
    ]
    formatter = InputOutputFormatter()
    algorithm = RepeReadingControl()
    training_data = algorithm._build_steering_vector_training_data(dataset, formatter)
    assert len(training_data) == 2
    assert "France" in training_data[0].positive_prompt
    assert "2" in training_data[1].positive_prompt


def test_RepeReadingControl_build_steering_vector_training_data_with_repeat_correct() -> (
    None
):
    dataset: Dataset = [
        Example(
            instruction="",
            input="Paris is in",
            output="France",
            incorrect_outputs=["Germany", "Italy"],
        ),
        Example(
            instruction="",
            input="1 + 1 =",
            output="2",
            incorrect_outputs=["11", "34", "3.14"],
        ),
    ]
    formatter = InputOutputFormatter()
    algorithm = RepeReadingControl(multi_answer_method="repeat_correct")
    training_data = algorithm._build_steering_vector_training_data(dataset, formatter)
    assert len(training_data) == 5
    # the positive example should be repeated once for each incorrect output
    assert "Germany" in training_data[0].negative_prompt
    assert "France" in training_data[0].positive_prompt
    assert "Italy" in training_data[1].negative_prompt
    assert "France" in training_data[1].positive_prompt
    assert "2" in training_data[2].positive_prompt
    assert "11" in training_data[2].negative_prompt
    assert "2" in training_data[3].positive_prompt
    assert "34" in training_data[3].negative_prompt
    assert "2" in training_data[4].positive_prompt
    assert "3.14" in training_data[4].negative_prompt


def test_RepeReadingControl_get_steering_vector(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    tokenizer.pad_token_id = model.config.eos_token_id
    pipeline = Pipeline(model, tokenizer)
    dataset: Dataset = [
        Example(
            instruction="",
            input="Paris is in",
            output="France",
            incorrect_outputs=["Germany", "Italy"],
        ),
    ]
    algorithm = RepeReadingControl(multi_answer_method="repeat_correct")
    steering_vector = algorithm._get_steering_vector(pipeline, dataset)
    assert list(steering_vector.layer_activations.keys()) == [0, 1, 2, 3, 4, 5]
    for act in steering_vector.layer_activations.values():
        assert act.shape == (512,)


def test_RepeReadingControl_get_steering_vector_matches_caa(
    empty_llama_model: LlamaForCausalLM, llama_chat_tokenizer: Tokenizer
) -> None:
    model = empty_llama_model
    tokenizer = llama_chat_tokenizer
    pipeline = Pipeline(
        model,
        tokenizer,
        prompter=LlamaChatPrompter(),
        formatter=LlamaChatFormatter(),
    )
    dataset: Dataset = [
        Example(
            instruction="",
            input="Paris is in",
            output="France",
            incorrect_outputs=["Germany", "Italy"],
        ),
    ]
    layers = [0, 1, 2]
    algorithm = RepeReadingControl(multi_answer_method="repeat_correct", layers=layers)
    steering_vector = algorithm._get_steering_vector(pipeline, dataset)

    steering_training_data = algorithm._build_steering_vector_training_data(
        dataset, pipeline.formatter
    )
    # hackily translated from generate_vectors.py script
    tokenized_data = [
        (tokenizer.encode(pos), tokenizer.encode(neg))
        for pos, neg in steering_training_data
    ]
    pos_activations = dict([(layer, []) for layer in layers])
    neg_activations = dict([(layer, []) for layer in layers])
    wrapped_model = LlamaWrapper(model, tokenizer)

    for p_tokens, n_tokens in tokenized_data:
        p_tokens = torch.tensor(p_tokens).unsqueeze(0).to(model.device)
        n_tokens = torch.tensor(n_tokens).unsqueeze(0).to(model.device)
        wrapped_model.reset_all()
        wrapped_model.get_logits(p_tokens)
        for layer in layers:
            p_activations = wrapped_model.get_last_activations(layer)
            p_activations = p_activations[0, -2, :].detach().cpu()
            pos_activations[layer].append(p_activations)
        wrapped_model.reset_all()
        wrapped_model.get_logits(n_tokens)
        for layer in layers:
            n_activations = wrapped_model.get_last_activations(layer)
            n_activations = n_activations[0, -2, :].detach().cpu()
            neg_activations[layer].append(n_activations)

    caa_vecs_by_layer = {}
    for layer in layers:
        all_pos_layer = torch.stack(pos_activations[layer])
        all_neg_layer = torch.stack(neg_activations[layer])
        caa_vecs_by_layer[layer] = (all_pos_layer - all_neg_layer).mean(dim=0)

    for layer in layers:
        assert torch.allclose(
            steering_vector.layer_activations[layer], caa_vecs_by_layer[layer]
        )


def test_RepeReadingControl_run(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    tokenizer.pad_token_id = model.config.eos_token_id
    pipeline = Pipeline(model, tokenizer)

    test_example = Example(
        instruction="",
        input="Paris is in",
        output="France",
        incorrect_outputs=["Germany", "Italy"],
    )
    dataset: Dataset = [
        test_example,
        Example(
            instruction="",
            input="1 + 1 =",
            output="2",
            incorrect_outputs=["11", "34", "3.14"],
        ),
    ]

    original_outputs = pipeline.generate(test_example)

    algorithm = RepeReadingControl(
        patch_generation_tokens_only=False, direction_multiplier=10
    )
    algorithm.run(pipeline, dataset)
    new_outputs = pipeline.generate(test_example)

    # TODO: find a better assertion that ensures this is actually doing what it should
    assert original_outputs != new_outputs


# TODO: uncomment this test when https://github.com/nrimsky/SycophancySteering/issues/2 is fixed
# def test_RepeReadingControl_run_steering_matches_caa_llama_wrapper(
#     empty_llama_model: LlamaForCausalLM, llama_chat_tokenizer: Tokenizer
# ) -> None:
#     model = empty_llama_model
#     tokenizer = llama_chat_tokenizer
#     pipeline = Pipeline(
#         model,
#         tokenizer,
#         prompter=LlamaChatPrompter(),
#         formatter=LlamaChatFormatter(),
#     )
#     test_example = Example(
#         instruction="",
#         input="Paris is in",
#         output="France",
#         incorrect_outputs=["Germany", "Italy"],
#     )
#     dataset: Dataset = [
#         test_example,
#         Example(
#             instruction="",
#             input="1 + 1 =",
#             output="2",
#             incorrect_outputs=["11", "34", "3.14"],
#         ),
#     ]

#     layers = [0, 1, 2]
#     multiplier = 7
#     algorithm = RepeReadingControl(
#         patch_generation_tokens_only=True,
#         direction_multiplier=multiplier,
#         layers=layers,
#     )
#     algorithm.run(pipeline, dataset)
#     hook = pipeline.hooks[0]

#     # hackily recreating what the pipeline does during logprobs
#     base_prompt = pipeline.build_generation_prompt(test_example)
#     full_prompt = base_prompt + test_example.output
#     inputs = tokenizer(full_prompt, return_tensors="pt").to(model.device)
#     ctx = PipelineContext(
#         method="logprobs",
#         base_prompt=base_prompt,
#         full_prompt=full_prompt,
#         inputs=inputs,
#         pipeline=pipeline,
#     )
#     orig_logits = model(**inputs).logits
#     with hook(ctx):
#         our_logits = model(**inputs).logits

#     assert isinstance(hook, SteeringHook)  # keep pyright happy
#     wrapped_model = LlamaWrapper(model, tokenizer, add_only_after_end_str=True)
#     wrapped_model.reset_all()
#     for layer in layers:
#         wrapped_model.set_add_activations(
#             layer, multiplier * hook.steering_vector.layer_activations[layer]
#         )
#     caa_logits = wrapped_model.get_logits(inputs["input_ids"])
#     # only the final answer tokens should be different
#     assert torch.allclose(our_logits[0, :-2], orig_logits[0, :-2])
#     assert torch.allclose(our_logits, caa_logits)


def test_RepeReadingControl_run_logprobs_with_patch_generation_tokens_only(
    model: GPTNeoXForCausalLM, tokenizer: Tokenizer
) -> None:
    tokenizer.pad_token_id = model.config.eos_token_id
    pipeline = Pipeline(model, tokenizer)

    test_example = Example(
        instruction="",
        input="Paris is in",
        output="France",
        incorrect_outputs=["Germany", "Italy"],
    )
    dataset: Dataset = [
        test_example,
        Example(
            instruction="",
            input="1 + 1 =",
            output="2",
            incorrect_outputs=["11", "34", "3.14"],
        ),
    ]

    original_outputs = pipeline.calculate_output_logprobs(test_example)

    algorithm = RepeReadingControl(patch_generation_tokens_only=True)
    algorithm.run(pipeline, dataset)
    new_outputs = pipeline.calculate_output_logprobs(test_example)

    assert original_outputs.sum_logprobs != new_outputs.sum_logprobs
    # only the final token should be patched and thus be different
    for old, new in zip(
        original_outputs.token_probs[:-1], new_outputs.token_probs[:-1]
    ):
        assert old.text == new.text
        assert old.logprob == new.logprob
    assert (
        original_outputs.token_probs[-1].logprob != new_outputs.token_probs[-1].logprob
    )


def test_find_generation_start_token_index_with_trailing_space(
    tokenizer: Tokenizer,
) -> None:
    base_prompt = "Paris is in: "
    full_prompt = "Paris is in: France"
    assert _find_generation_start_token_index(tokenizer, base_prompt, full_prompt) == 3


def test_find_generation_start_token_index_with_trailing_special_chars(
    tokenizer: Tokenizer,
) -> None:
    base_prompt = "<s> Paris is in: </s>"
    full_prompt = "<s> Paris is in: France </s>"
    assert _find_generation_start_token_index(tokenizer, base_prompt, full_prompt) == 6


def test_find_generation_start_token_base(
    tokenizer: Tokenizer,
) -> None:
    base_prompt = "<s> Paris is in:"
    full_prompt = "<s> Paris is in: France"
    assert _find_generation_start_token_index(tokenizer, base_prompt, full_prompt) == 6
