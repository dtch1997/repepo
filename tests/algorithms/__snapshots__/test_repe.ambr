# serializer version: 1
# name: test_RepeReadingControl_build_steering_vector_training_data_picks_one_neg_by_default
  list([
    SteeringVectorTrainingSample(
      negative_prompt='''
        <s> Input:  Paris is in 
        Output: Germany
      ''',
      positive_prompt='''
        <s> Input:  Paris is in 
        Output: France
      ''',
    ),
    SteeringVectorTrainingSample(
      negative_prompt='''
        <s> Input:  1 + 1 = 
        Output: 11
      ''',
      positive_prompt='''
        <s> Input:  1 + 1 = 
        Output: 2
      ''',
    ),
  ])
# ---
